File: python/GC/garbage_allocator.py
import const
import GC.garbage_holder as gh
import time
import datetime 
import common.interval

class GarbageAllocator:
    def __init__(self, 
        RunDuration, 
        Allocations, 
        StartIndex, 
        GarbageHolder,
        AllocationCount,
        ByteCount,
        GCPauses,
        EndTimestamp):
        self.RunDuration = RunDuration
        self.Allocations = Allocations
        self.StartIndex = StartIndex
        self.GarbageHolder = GarbageHolder
        self.AllocationCount = AllocationCount
        self.ByteCount = ByteCount
        self.GCPauses = GCPauses
        self.EndTimestamp = EndTimestamp

# Since some parameters are fixed and others 

def NewGabageAllocator(runDuration, allocations, startIndex):
    return GarbageAllocator(runDuration, 
    allocations, 
    startIndex, 

    gh.GarbageHolder())

class AllocationInfo:
    def __init__(self,
    ArraySize,
    BucketIndex,
    Generation):
        self.ArraySize = ArraySize
        self.BucketIndex = BucketIndex
        self.Generation = Generation



class AllocationInfo:
    def __init__(self, ArraySize, BucketIndex, GenerationIndex):
        self.ArraySize = ArraySize
        self.BucketIndex = BucketIndex
        self.GenerationIndex = GenerationIndex

def create_garbage(arraySize):
    # Here we return an array of size arraySize
    return [0]*arraySize
def run_garbage_allocator(GarbageAllocator):
    allocations = GarbageAllocator.Allocations
    gh = GarbageAllocator.GarbageHolder
    gcPauses = GarbageAllocator.GCPauses
    allocationIndex = GarbageAllocator.StartIndex
    allocationIndexMask = len(allocations)- 1
    gh.Start(GarbageAllocator.GarbageHolder)

    lastTimestamp = time.time_ns()
    endTimestamp = lastTimestamp + GarbageAllocator.RunDuration
    while(lastTimestamp < endTimestamp):
        allocation_index = allocations[allocationIndex]
        allocationIndex = (allocationIndex+1) & allocationIndexMask

        # Allocation
        garbage = gh.CreateGarbage(allocation_index.ArraySize)
        gh.AddGarbage(garbage, allocation_index.BucketIndex, allocation_index.GenerationIndex)
        GarbageAllocator.AllocationCount = GarbageAllocator.AllocationCount +1
        GarbageAllocator.ByteCount = GarbageAllocator.ByteCount

        # Release %15 content
        if(allocationIndex % 15) == 10:
            gh.TryReleaseGarbage()

        # GC check
        # This elapsed time should be in nanoseconds
        elapsed = time.time()
        diff = elapsed - lastTimestamp
        if diff >= const.MinGCPause:
            GCPauses = gcPauses.append(common.interval.Interval(lastTimestamp,elapsed))
        lastTimestamp = elapsed
    
    GarbageAllocator.EndTimestamp = datetime.datetime.now()
    GarbageAllocator.GCPauses = GCPauses

def ByteSizeToArraySize(byteSize):
    s = (byteSize - const.MinAllocationSize+const.ArrayItemSize - 1) /const.ArrayItemSize
    if s < 0:
        return 0
    return s

def ArraySizeToByteSize(arraySize):
    return const.MinAllocationSize + arraySize << const.ArrayItemSizeLog2

File: Go/burn_test/garbage_allocator.go
package burn_test

import (
	. "../common"
	. "../common/nanotime"
	"time"
)

const AvgPauseFrequency = 12000

type GarbageAllocator struct {
	RunDuration   time.Duration
	Allocations   []AllocationInfo
	StartIndex    int32
	GarbageHolder *GarbageHolder

	// Statistics
	AllocationCount int64
	ByteCount       int64
	GCPauses        []Interval
	EndTimestamp    int64
}

type AllocationInfo struct {
	ArraySize       int32
	BucketIndex     int8
	GenerationIndex int8
}

func NewGarbageAllocator(runDuration time.Duration, allocations []AllocationInfo, startIndex int32) *GarbageAllocator {
	a := &GarbageAllocator{
		RunDuration:   runDuration,
		Allocations:   allocations,
		StartIndex:    startIndex,
		GarbageHolder: NewGarbageHolder(),
		GCPauses:      make([]Interval, 0, int32(runDuration.Seconds()*AvgPauseFrequency)), // No reallocations during the test
	}
	return a
}

func ByteSizeToArraySize(byteSize int32) int32 {
	s := (byteSize - MinAllocationSize + ArrayItemSize - 1) / ArrayItemSize
	if s < 0 {
		return 0
	}
	return s
}

func ArraySizeToByteSize(arraySize int32) int64 {
	return int64(MinAllocationSize) + (int64(arraySize) << ArrayItemSizeLog2)
}

func CreateGarbage(arraySize int32) []int64 {
	//make function usage -> https://go.dev/tour/moretypes/13
	return make([]int64, arraySize)
}

func (a *GarbageAllocator) Run() {
	allocations := a.Allocations
	gh := a.GarbageHolder
	gcPauses := a.GCPauses

	allocationIndex := a.StartIndex
	allocationIndexMask := int32(len(allocations) - 1)

	a.GarbageHolder.Start()
	lastTimestamp := Nanotime().Nanoseconds()
	endTimestamp := lastTimestamp + a.RunDuration.Nanoseconds()

	for lastTimestamp < endTimestamp {
		ai := allocations[allocationIndex]
		allocationIndex = (allocationIndex + 1) & allocationIndexMask

		// Allocation
		var garbage = CreateGarbage(ai.ArraySize)
		gh.AddGarbage(garbage, ai.BucketIndex, ai.GenerationIndex)

		a.AllocationCount++
		a.ByteCount += ArraySizeToByteSize(ai.ArraySize)

		// Releasing what should be released by now
		if (allocationIndex & 15) == 0 {
			gh.TryReleaseGarbage()
		}

		// GC check
		elapsed := Nanotime().Nanoseconds()
		diff := elapsed - lastTimestamp
		if diff >= MinGCPause {
			gcPauses = append(gcPauses, Interval{
				Start: float64(lastTimestamp),
				End:   float64(elapsed),
			})
		}
		lastTimestamp = elapsed
	}
	a.EndTimestamp = Nanotime().Nanoseconds()
	a.GCPauses = gcPauses
}
